cpp/src/arrow/compute/exec.cc

```diff
@@ -498,15 +510,36 @@ struct NullGeneralization {
     return PERHAPS_NULL;
   }
 
+  static type Get(const ChunkedArray& chunk_array) {
+    if (chunk_array.num_chunks() == 0 ||
+        (chunk_array.null_count() == chunk_array.length())) {
+      return ALL_NULL;
+    }
+
+    type null_gen = ALL_NULL;
+    int chunk_idx = 0;
+    while (chunk_idx < chunk_array.num_chunks()) {
+      ExecValue value;
+      const ArrayData* curr_chunk = chunk_array.chunk(chunk_idx)->data().get();
+      value.SetArray(*curr_chunk);
+      null_gen = static_cast<type>((null_gen & Get(value)));
```

Comment:
it's better to do the switch on null_gen here than to rely on `&` and clever number assignment to the enum

Reply:
if you get a `PERHAPS_NULL` you can early return

Reply:
I've thought about this, you're right, we shouldn't rely on `&`, it's too tricky and there will be potential scalability issues.

Reply:
Personally I think LLVM/gcc is smart enough and this might not a bottleneck...

Reply:
> Personally I think LLVM/gcc is smart enough and this might not a bottleneck...

There is a potential problem here. The previous calculation logic will rely on ALL_NULL = 3, so the null_gen check for all chunks in a chunk-array can be done with `&`.
For example: chunk-array returns `ALL_VALID` only if all chunks are `ALL_VALID`. The calculation through `&` must rely on `ALL_NULL = 3`, which is actually unreasonable.

Reply:
> The previous calculation logic will rely on ALL_NULL = 3, so the null_gen check for all chunks in a chunk-array can be done with &.

I mean we shouldn't rely on that because this is even not a bottleneck, if it's maybe we should extract a wrapper for this rather than bithack the enum, and let LLVM / gcc do further opt

Reply:
Yes, you're right, GCC/LLVM will optimize enum type calculations as much as possible.